Python/peephole.c

```diff
@@ -536,3 +527,45 @@ PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
     PyMem_Free(codestr);
     return code;
 }
+
+/* Perform basic peephole optimizations to components of a code object.
+   The consts object should still be in list form to allow new constants
+   to be appended.
+
+   To keep the optimizer simple, it bails when the lineno table has complex
+   encoding for gaps >= 255.
+
+   Optimizations are restricted to simple transformations occurring within a
+   single basic block.  All transformations keep the code size the same or
+   smaller.  For those that reduce size, the gaps are initially filled with
+   NOPs.  Later those NOPs are removed and the jump addresses retargeted in
+   a single pass. */
+
+PyObject *
+PyCode_Optimize(PyObject *code, PyObject* consts, PyObject *names,
+                PyObject *lnotab_obj)
+{
+    PyObject* bytecode = NULL;
+    PyObject* old_bytecode = code;
+    int compare_bytecode;
+    do {
+        bytecode = old_bytecode;
+        bytecode = optimize_bytecode_once(bytecode, consts, names, lnotab_obj);
+        if (!bytecode) {
+            if (old_bytecode != code) {
+                Py_DECREF(old_bytecode);
+            }
+            return NULL;
+        }
+        compare_bytecode = PyObject_RichCompareBool(old_bytecode, bytecode, Py_NE);
+        if (old_bytecode != code) {
+            Py_DECREF(old_bytecode);
+        }
+        if (compare_bytecode == -1) {
+            break;
```

Comment:
I don't think the `PyObject_RichCompareBool()` call is being handled properly. A return value of `-1` means that an error occurred, so `PyCode_Optimize()` should return `NULL` in that case.